src/target/target_flash.c

```diff
@@ -0,0 +1,271 @@
+/*
+ * This file is part of the Black Magic Debug project.
+ *
+ * Copyright (C) 2022  Black Sphere Technologies Ltd.
+ * Copyright (C) 2022  Rafael Silva <perigoso@riseup.net>
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/* This file implements target flash interaction routines.
+ * Provides functionality for buffered flash operations
+ * It depends on target flash implementations
+ */
+
+#include "general.h"
+#include "target_internal.h"
+
+static int flash_buffered_write(target_flash_s *f, target_addr_t dest, const void *src, size_t len);
+static int flash_buffered_flush(target_flash_s *f);
+
+target_flash_s *target_flash_for_addr(target *t, uint32_t addr)
+{
+	for (target_flash_s *f = t->flash; f; f = f->next)
+		if ((f->start <= addr) && (addr < (f->start + f->length)))
+			return f;
+	return NULL;
+}
+
+static int target_enter_flash_mode(target *t)
+{
+	if (t->flash_mode)
+		return 0;
+
+	int ret = 0;
+	if (t->enter_flash_mode)
+		ret = t->enter_flash_mode(t);
+	else
+		/* Reset target on flash command */
+		/* This saves us if we're interrupted in IRQ context */
+		target_reset(t);
+
+	if (ret == 0)
+		t->flash_mode = true;
+
+	return ret;
+}
+
+static int target_exit_flash_mode(target *t)
+{
+	if (!t->flash_mode)
+		return 0;
+
+	int ret = 0;
+	if (t->exit_flash_mode)
+		ret = t->exit_flash_mode(t);
+	else
+		/* Reset target to known state when done flashing */
+		target_reset(t);
+
+	t->flash_mode = false;
+
+	return ret;
+}
+
+static int flash_prepare(target_flash_s *f)
+{
+	if (f->ready)
+		return 0;
+
+	int ret = 0;
+	if (f->prepare)
+		ret = f->prepare(f);
+
+	if (ret == 0)
+		f->ready = true;
+
+	return ret;
+}
+
+static int flash_done(target_flash_s *f)
+{
+	if (!f->ready)
+		return 0;
+
+	int ret = 0;
+	if (f->done)
+		ret = f->done(f);
+
+	if (f->buf) {
+		free(f->buf);
+		f->buf = NULL;
+	}
+
+	f->ready = false;
+
+	return ret;
+}
+
+int target_flash_erase(target *t, target_addr_t addr, size_t len)
+{
+	if (target_enter_flash_mode(t) != 0)
+		return 1;
+
+	int ret = 0;
+	while (len) {
```

Comment:
Would it be possible to pivot this into a for loop (eg, `for (size_t offset = 0; offset < len; offset += f->blocksize)`)? It might help with code flow and simplify the shenanigans at the end of the loop

Reply:
I will look into it, I do generally prefer for loops

Reply:
here I don't think so, as we don't actually know what flashes we are going to work with, and thus no block size, it's also before alignment, which depends on the block size too

Reply:
Think it'd be ok if the line immediately below was lifted out the loop, but it's something to play with later. For now, we'll accept the explanation and run with this as it's just such a big improvement anyway that refactoring and improvements can come later.